### PR TITLE
UCS/CONFIG: Add tests for unused env vars functionality

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

Add tests for unused env vars functionality

## Why ?

To safely continue further development/maintenance that touches related code 

## How ?

1. Move `warn_once` flag to the caller's code (i.e. UCP's code)
2. Add UCX environment cleanup class that:
- ctor: `unsetenv` all UCX env vars
- dtor: restore all UCX env vars
3. Add the tests

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10320/ for details (Mellanox internal link).

File: src/ucp/core/ucp_worker.c [hunk 1]
```diff
@@ -1462,6 +1462,17 @@ unsigned ucp_worker_get_ep_config(ucp_worker_h worker,
     return config_idx;
 }
 
+static void ucp_worker_warn_config_unused_env_vars()
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: can we move this to ucs? so gtest will call internal function to do the testing and make sure warning is printed, but ucp_worker_create will still be calling ucs_config_parser_warn_unused_env_vars() like before

File: test/gtest/common/test_helpers.cc [hunk 2]
```diff
@@ -276,6 +278,40 @@ scoped_setenv::~scoped_setenv() {
...
+ucx_env_cleanup::ucx_env_cleanup() {
+    const size_t prefix_len = strlen(UCS_CONFIG_PREFIX);
+    const char *var_name, *var_value;
+    char **envp, *saveptr;
+
+    for (envp = environ; *envp != NULL; ++envp) {
+        std::string env_var = *envp;
+
+        var_name = strtok_r(const_cast<char*>(env_var.c_str()), "=", &saveptr);
+        if (!var_name || strncmp(var_name, UCS_CONFIG_PREFIX, prefix_len)) {
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc]: this is wrong to const-cast here, std::string returns const char* for a reason

Comment (dmitrygladkov) [reply test/gtest/common/test_helpers.cc]: fixed - used the same approach as in the ucs/config/parser

File: src/ucp/core/ucp_worker.c (hunk 1 above)
Comment (dmitrygladkov) [reply src/ucp/core/ucp_worker.c]: sure, we can do it - done

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10321/ for details (Mellanox internal link).

File: test/gtest/common/test_helpers.cc [hunk 3]
```diff
@@ -276,6 +278,43 @@ scoped_setenv::~scoped_setenv() {
...
+    for (envp = environ; *envp != NULL; ++envp) {
+        envstr = ucs_strdup(*envp, "env_str");;
+
+        var_name = strtok_r(envstr, "=", &saveptr);
+        if (!var_name || strncmp(var_name, UCS_CONFIG_PREFIX, prefix_len)) {
+            ucs_free(envstr);
+            continue; /* Not UCX */
+        }
+
+        var_value = strtok_r(NULL, "=", &saveptr);
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc]: why do we search for another '=' char? we should take the remainder of the string as-is
consider env var value with '=' inside

File: test/gtest/common/test_helpers.cc [hunk 4]
```diff
@@ -276,6 +278,43 @@ scoped_setenv::~scoped_setenv() {
...
+        if (!var_value) {
+            var_value = "";
+        }
+
+        ucx_env_storage.push_back(std::make_pair(var_name, var_value));
+
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc]: redundant space line

File: test/gtest/common/test_helpers.cc (hunk 3 above)
Comment (dmitrygladkov) [reply test/gtest/common/test_helpers.cc]: @yosefe do we need this to call `putenv` instead? if yes, I've done it

File: test/gtest/common/test_helpers.cc (hunk 4 above)
Comment (dmitrygladkov) [reply test/gtest/common/test_helpers.cc]: fixed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10322/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10323/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10324/ for details (Mellanox internal link).

File: test/gtest/common/test_helpers.cc [hunk 5]
```diff
@@ -287,25 +287,27 @@ ucx_env_cleanup::ucx_env_cleanup() {
...
 ucx_env_cleanup::~ucx_env_cleanup() {
     while (!ucx_env_storage.empty()) {
-        char *env_var = ucx_env_storage.back();
+        std::string var_name =
+            ucx_env_storage.back().substr(0, ucx_env_storage.back().find("="));
+        std::string var_value =
+            ucx_env_storage.back().substr(ucx_env_storage.back().find("=") + 1);
 
-        putenv(ucx_env_storage.back());
+        setenv(var_name.c_str(), var_value.c_str(), 1);
```
Comment (yosefe) [inline test/gtest/common/test_helpers.cc]: why not using putenv?
what if '=' is not found in the string?

Comment (dmitrygladkov) [reply test/gtest/common/test_helpers.cc]: > why not using putenv?

`putenv` requires to keep a string buffer passed to the function untouchable and then `free` this memory.

> what if '=' is not found in the string?

good question, such elements shouldn't be processed - we have to skip them

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/10337/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@brminich could you take a look pls?

**Pull request merged**